stdlib/internal/SwiftSyntax/Syntax.swift

```diff
@@ -0,0 +1,183 @@
+  ///         be preserved in all circumstances where Syntax nodes are created.
+  internal unowned var data: SyntaxData
+
+#if DEBUG
+  func validate() {
+    // This is for subclasses to override to perform structural validation.
+  }
+#endif
+
+  /// Creates a Syntax node from the provided root and data.
+  internal init(root: SyntaxData, data: SyntaxData) {
+    self._root = root
+    self.data = data
+#if DEBUG
+    validate()
+#endif
+  }
+
+  /// Access the raw syntax assuming the node is a Syntax.
+  var raw: RawSyntax {
+    return data.raw
+  }
+
+  /// An iterator over children of this node.
+  public var children: SyntaxChildren {
+    return SyntaxChildren(node: self)
+  }
+
+  /// Whether or not this node it marked as `present`.
+  public var isPresent: Bool {
+    return raw.presence == .present
+  }
+
+  /// Whether or not this node it marked as `missing`.
+  public var isMissing: Bool {
+    return raw.presence == .missing
+  }
+
+  /// Whether or not this node represents an Expression.
+  public var isExpr: Bool {
+    return raw.kind.isExpr
+  }
+  
+  /// Whether or not this node represents a Declaration.
+  public var isDecl: Bool {
+    return raw.kind.isDecl
+  }
+
+  /// Whether or not this node represents a Statement.
+  public var isStmt: Bool {
+    return raw.kind.isStmt
+  }
+
+  /// Whether or not this node represents a Type.
+  public var isType: Bool {
+    return raw.kind.isType
+  }
+
+  /// Whether or not this node represents a Pattern.
+  public var isPattern: Bool {
+    return raw.kind.isPattern
+  }
+
+  /// The parent of this syntax node, or `nil` if this node is the root.
+  public var parent: Syntax? {
+    guard let parentData = data.parent else { return nil }
+    return Syntax.make(root: _root, data: parentData)
+  }
+
+  /// The root of the tree in which this node resides.
+  public var root: Syntax {
+    return Syntax.make(root: _root,  data: _root)
+  }
+
+  /// Prints the raw value of this node to the provided stream.
+  /// - Parameter stream: The stream to which to print the raw tree.
+  func print<StreamType: TextOutputStream>(to stream: inout StreamType) {
+    data.raw.print(to: &stream)
+  }
+  
+  /// Gets the child at the provided index in this node's children.
+  /// - Parameter index: The index of the child node you're looking for.
+  /// - Returns: A Syntax node for the provided child, or `nil` if there
+  ///            is not a child at that index in the node.
+  public func child(at index: Int) -> Syntax? {
+    guard raw.layout.indices.contains(index) else { return nil }
+    if raw.layout[index].isMissing { return nil }
+    return Syntax.make(root: _root, data: data.cachedChild(at: index))
+  }
+
+  /// A source-accurate description of this node.
+  public var description: String {
+    var s = ""
+    self.print(to: &s)
+    return s
+  }
+}
+
+/// Determines if two nodes are equal to each other.
+public func ==<T: Syntax>(lhs: T, rhs: T) -> Bool {
```

Comment:
Ah! Yeah, I had originally structured this with protocols, and didn't want to enforce Equatable conformance because of the Self requirement, but since this is a class hierarchy I can make this a conformance.